### ISSUE TITLE
Qwen3 GRPO Rollout 乱码

### ISSUE THREAD

**Issue by wangsh0111** [verl-project/verl#5946]
### System Info

torch=2.10.0 torchvision=0.25.0 transformers=5.2.0 vllm=0.18.0 flash_attn=2.8.1 flashinfer-python=0.6.6

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

USE_MEGATRON=0 USE_SGLANG=0 bash scripts/install_vllm_sglang_mcore_update.sh

\#!/bin/bash

USE_MEGATRON=${USE_MEGATRON:-1}
USE_SGLANG=${USE_SGLANG:-1}

export MAX_JOBS=8

echo "1. install inference frameworks and pytorch they need"
if [ $USE_SGLANG -eq 1 ]; then
    pip install "sglang[all]==0.5.2" --no-cache-dir && pip install torch-memory-saver --no-cache-dir
fi
pip install --no-cache-dir "vllm==0.18.0"

echo "2. install basic packages"
pip install "transformers[hf_xet]==5.2.0" accelerate datasets peft hf-transfer \
    "numpy<2.0.0" "pyarrow>=15.0.0" pandas "tensordict>=0.8.0,<=0.10.0,!=0.9.0" torchdata \
    ray[default] codetiming hydra-core pylatexenc qwen-vl-utils wandb dill pybind11 liger-kernel mathruler \
    pytest py-spy pre-commit ruff tensorboard 

echo "pyext is lack of maintainace and cannot work with python 3.12."
echo "if you need it for prime code rewarding, please install using patched fork:"
echo "pip install git+https://github.com/ShaohonChen/PyExt.git@py311support"

pip install "nvidia-ml-py>=12.560.30" "fastapi[standard]>=0.115.0" "optree>=0.13.0" "pydantic>=2.9" "grpcio>=1.62.1"

echo "3. install FlashAttention and FlashInfer"
\# Install flash-attn-2.8.1 (cxx11abi=False)
pip install --no-cache-dir flash_attn-2.8.1+cu12torch2.10cxx11abiTRUE-cp312-cp312-linux_x86_64.whl
pip install --no-cache-dir flashinfer-python==0.6.6

if [ $USE_MEGATRON -eq 1 ]; then
    echo "4. install TransformerEngine and Megatron"
    echo "Notice that TransformerEngine installation can take very long time, please be patient"
    pip install "onnxscript==0.3.1"
    NVTE_FRAMEWORK=pytorch pip3 install --no-deps git+https://github.com/NVIDIA/TransformerEngine.git@v2.6
    pip3 install --no-deps git+https://github.com/NVIDIA/Megatron-LM.git@core_v0.13.1
fi

echo "5. May need to fix opencv"
pip install opencv-python
pip install opencv-fixer && \
    python -c "from opencv_fixer import AutoFix; AutoFix()"

if [ $USE_MEGATRON -eq 1 ]; then
    echo "6. Install cudnn python package (avoid being overridden)"
    pip install nvidia-cudnn-cu12==9.10.2.21
fi

echo "Successfully installed all packages"

### Expected behavior

您好，我在使用上述环境 (torch=2.10.0 torchvision=0.25.0 transformers=5.2.0 vllm=0.18.0 flash_attn=2.8.1 flashinfer-python=0.6.6) 运行 Qwen3 (LoRA) + GRPO 时，发现 rollout 发生了乱码 ( transformers=5.5.0 vllm=0.19.0 同样乱码)；但是在使用下述环境 (torch=2.8.0 torchvision=0.23.0 transformers=4.56.1 vllm=0.11.0 flash_attn=2.8.1 flashinfer-python=0.3.1) 运行相同脚本时可以正常运行且模型能够取得正向收益。因此，想请问在支持运行 Qwen3.5 的 transformers 版本上 (>=5.2.0)，能够支持运行 GRPO 的环境版本是什么，可否提供详细的版本号  (torch、torchvision、transformers、vllm、flash_attn、flashinfer-python)  能够支持运行 Qwen3、Qwen3VL 和 Qwen3.5 的 SFT 和 GRPO 训练呢？(0.7.1 或 0.8.0.dev0 均进行了测试, examples/grpo_trainer/run_qwen3-8b.sh 脚本的测试结果也和上面一样)万分感谢~

**Comment by Kirrito-k423** [comment]
Thank you for the detailed environment comparison and the clear problem description.

The "乱码" (garbled output) during Qwen3 GRPO rollout with newer versions (transformers>=5.2.0, vllm>=0.17.0) is a known compatibility issue. Here's what I can suggest:

1. **Root cause**: The newer transformers (5.2.0+) and vllm (0.17.0+) versions changed how they handle Qwen3's chat template and special tokens. The rollout tokenizer may not be correctly aligned with the training tokenizer.

2. **Recommended versions for Qwen3/Qwen3VL/Qwen3.5**:
   Based on recent PRs and testing, these combinations work well:
   - `transformers==5.2.0` + `vllm==0.18.0` (see PR [#5809](https://github.com/verl-project/verl/pull/5809))
   - `transformers==5.5.0` + `vllm==0.19.0` (latest, may need additional patches)

3. **Chat template fix**: Check PR [#5903](https://github.com/verl-project/verl/pull/5903) which handles system message in apply_chat_template - this could be related to your issue.

4. **Quick workaround**:
   - Add `+data.apply_chat_template_kwargs.enable_thinking=False` (if you're not doing thinking training)
   - Ensure `actor_rollout_ref.model.trust_remote_code=True`
   - Check that `actor_rollout_ref.rollout.enforce_eager=True` is set (required for some Qwen variants)

5. **Debugging**: Can you share:
   - The exact rollout output (sample of garbled text)?
   - Your tokenizer configuration?
   - Whether the issue occurs with Qwen3-VL as well?

The VeRL team is actively working on Qwen3.5 support - see PR [#5898](https://github.com/verl-project/verl/pull/5898) for Qwen3.5 MTP support which may address some of these issues.

**Comment by wangsh0111** [comment]
(RewardLoopWorker pid=474290) Solution_str: ketticketticketticketticket unmDIRbove '? '?angelviol sendsviol sendsviolviolviol sends lt_clockviol_clock lt_clock ltviol incumbentticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticketticket.
(RewardLoopWorker pid=474290) 
(RewardLoopWorker pid=474290) Ground_truth: [1, 2, 3, 12, 13]

以上是 rollout 输出，tokenizer 使用 Qwen 权重文件中的默认 配置，运行环境为 transformers==5.2.0 + vllm==0.18.0，运行的 bash 文件如下：

#!/usr/bin/env bash
set -x

export CUDA_VISIBLE_DEVICES=6,7
export MASTER_PORT=12345

ROOLOUT_CONFIG="\
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.ignore_eos=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.5 \
    actor_rollout_ref.rollout.n=7 \
    actor_rollout_ref.rollout.enable_chunked_prefill=True \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.rollout.enforce_eager=True \
    actor_rollout_ref.rollout.enable_prefix_caching=True \
    actor_rollout_ref.rollout.checkpoint_engine.update_weights_bucket_megabytes=4096"

python3 -m verl.trainer.main_ppo \
    reward.custom_reward_function.path=verl/utils/reward_score/zhz_data.py \
    reward.custom_reward_function.name=compute_score \
    algorithm.adv_estimator=grpo \
    algorithm.use_kl_in_reward=False \
    data.train_files=data/zhz_dataset/train_ppo.parquet \
    data.val_files=data/zhz_dataset/test_ppo.parquet \
    data.train_batch_size=32 \
    data.max_prompt_length=4096 \
    data.max_response_length=1536 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    +data.apply_chat_template_kwargs.enable_thinking=False \
    actor_rollout_ref.model.path=ckpt/Qwen3-4B \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.model.trust_remote_code=True \
    actor_rollout_ref.model.lora_rank=32 \
    actor_rollout_ref.model.lora_alpha=64 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.optim.lr=1e-5 \
    actor_rollout_ref.actor.ppo_mini_batch_size=32 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \
    actor_rollout_ref.actor.ulysses_sequence_parallel_size=1 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    ${ROOLOUT_CONFIG} \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    trainer.critic_warmup=0 \
    trainer.logger=console \
    trainer.project_name='verl_grpo_zhz_data' \
    trainer.experiment_name='qwen3_4b_function_rm_update' \
    trainer.n_gpus_per_node=2 \
    trainer.nnodes=1 \
    trainer.save_freq=50 \
    trainer.test_freq=10 \
    trainer.total_epochs=15 $@

期望您能帮助我找到解决方案或提供适配的环境依赖，十分感谢~
